tortoise/backends/aioodbc/client.py

```diff
@@ -0,0 +1,277 @@
+import asyncio
+from functools import wraps
+from typing import Any, Callable, List, Optional, SupportsInt, Tuple, TypeVar, Union
+
+import aioodbc
+import pyodbc
+# from asyncpg.transaction import Transaction
+import pypika
+from pypika import OracleQuery as OrigOracleQuery
+
+
+from tortoise.backends.aioodbc.executor import AioodbcExecutor
+from tortoise.backends.aioodbc.schema_generator import AioodbcSchemaGenerator
+from tortoise.backends.base.client import (
+    BaseDBAsyncClient,
+    BaseTransactionWrapper,
+    Capabilities,
+    ConnectionWrapper,
+    NestedTransactionPooledContext,
+    PoolConnectionWrapper,
+    TransactionContext,
+    TransactionContextPooled,
+)
+from tortoise.exceptions import (
+    DBConnectionError,
+    IntegrityError,
+    OperationalError,
+    TransactionManagementError,
+)
+
+FuncType = Callable[..., Any]
+F = TypeVar("F", bound=FuncType)
+
+
+def translate_exceptions(func: F) -> F:
+    @wraps(func)
+    async def translate_exceptions_(self, *args):
+        try:
+            return await func(self, *args)
+        except pyodbc.ProgrammingError as exc:
+            raise OperationalError(exc)
+        except pyodbc.IntegrityError as exc:
+            raise IntegrityError(exc)
+        except pyodbc.Error as exc:
+            raise OperationalError(exc)
+        # except pyodbc.erras exc:  # pragma: nocoverage
+            # raise TransactionManagementError(exc)
+
+    return translate_exceptions_  # type: ignore
+
+
+
+
+
+
+from pypika import OracleQuery as OrigOracleQuery
+from pypika.queries import QueryBuilder, Query
+from pypika import Dialects
+
+
+class OracleQueryBuilder(QueryBuilder):
+    def __init__(self, **kwargs):
+        super(OracleQueryBuilder, self).__init__(dialect=Dialects.ORACLE, **kwargs)
+
+    def get_sql(self, *args, **kwargs):
+        return super(OracleQueryBuilder, self).get_sql(
+            *args, groupby_alias=False, **kwargs
+        )
+
+    def _limit_sql(self):
+        return f' FETCH NEXT {self._limit} ROWS ONLY'
+
+
+class OracleQuery(Query):
+    """
+    Defines a query class for use with Oracle.
+    """
+
+    @classmethod
+    def _builder(cls, **kwargs):
+        builder = OracleQueryBuilder(**kwargs)
+        builder.QUOTE_CHAR = ""
+        return builder
```

Comment:
How much of this is Oracle specific, and how much is aioodbc specific? Should we consider upstreaming this?

Reply:
The quote char and the limit workarounds are Oracle Specific.

Reply:
Ok, so at some point we try and upstream this. But only once its all up and running right.

Reply:
I see what you are asking now. See comments re-pypika upstream below. I will still check in with them to ask if they will accept my changes. Upstreaming would be ideal.